README.md
# KPI
You can use it only inside private VPN and only after authorization.

Depends:
```
python-qt4
python3-pyqt4
```
Optional:
```
python-qt4-dev
pyqt4-dev-tools
```
Debian package can be found in branch with the same name. More information: https://github.com/ink-ru/KPI/blob/debian/opt/kpi/README.md


DEBIAN/changelog
PACKAGE (VERSION) UNRELEASED; urgency=medium

  * Initial release.

 -- Васин Юрий <george.wise@gmail.com>  Mon, 22 Aug 2016 20:29:34 +0300


kpi_dicts.py
# -*- coding: utf-8 -*-

domain_url = 'aHR0cDovL290cC5kZW1pcy5ydQ=='
api_uri = "/smoke/oto/uto-kpi-api.php"
kpi_uri = "/smoke/oto/uto-kpi-tmp.php"
api_employees_get = "?action=users&da-rest=json"
api_result_get = "?action=result&da-rest=json"

result_rus_dict = {
'labor':				'Объем набранных баллов',
'labor_vip':			'Объем баллов за проверки VIP',
'issues_cnt':			'Количество выполненных задач',
'idle_penalty':			'Штрафные баллы за просрочку задач',
'overdue_cnt':			'Количество просроченных задач',
'result':				'Итоговое число баллов',
'intime':				'Количество задач выполненных в срок',
'intime_perc':			'Процент задач выполненных в срок',
'plan':					'План',
'plan_amnesty':			'Порог амнистии',
'absence':				'Отсутствия',
'plan_perc':			'Процент плана',
'money':				'Премия (итог)',
'money_result':			'Премия (факт)',
'vacation_labor':		'Отсутствия: компенсация',
'dept_labor':			'Подразделение: Объем набранных баллов',
'dept_labor_vip':		'Подразделение: Объем баллов за проверки VIP',
'dept_issues_cnt':		'Подразделение: Количество выполненных задач',
'dept_idle_penalty':	'Подразделение: Штрафные баллы за просрочку задач',
'dept_overdue_cnt':		'Подразделение: Количество просроченных задач',
'dept_result':			'Подразделение: Итоговое число баллов',
'dept_intime':			'Подразделение: Количество задач выполненных в срок',
'dept_intime_perc':		'Подразделение: Процент задач выполненных в срок',
'dept_plan':			'Подразделение: План',
'dept_plan_perc':		'Подразделение: Процент плана',
'dept_money_labor':		'Подразделение: Премия',
'dept_money_intime':	'Подразделение: Премия за сроки',
'forecast_ratio':		'Прогноз: Базовое отношение',
'labor_forecast':		'Прогноз: Объем набранных баллов',
'labor_vip_forecast':	'Прогноз: Объем баллов за проверки VIP',
'idle_penalty_forecast':'Прогноз: Штрафные баллы',
'result_forecast':		'Прогноз: Итоговое число баллов',
'plan_perc_forecast':	'Прогноз: Процент плана',
'money_forecast':		'Прогноз: Премия (итог)',
'money_result_forecast':'Прогноз: Премия (факт)',
'dept_labor_forecast':	'Прогноз подразделение: Баллы',
'dept_labor_vip_forecast':'Прогноз подразделение: Баллы за VIP',
'dept_result_forecast':	'Прогноз подразделение: Итоговое число баллов',
'dept_plan_perc_forecast':'Прогноз подразделение: Процент плана',
'dept_money_labor_forecast':'Прогноз подразделение: Премия',
# generatd keys
'1_labor':'баллы чистые',
'2_result':'баллы грязные',
# '':'',
}

plan_rate_dict = {
'Ведущий программист':450,
'Старший программист':400,
'Программист':350,
'Стажер-1М':87.5, # испытательный срок
'Стажер-1М':210,
'Стажер-3М':315,
'Исп-1М':100, # повышенный грейд
'Исп-1М':240,
'Исп-3М':360
}


opt/kpi/kpi.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import sys, re, urllib, json, collections, base64
import urllib.parse
import urllib.request
import operator # http://stackoverflow.com/questions/19411101/pyside-qtableview-example
import subprocess

from PyQt4.QtCore import QSettings
from PyQt4.QtCore import *
from PyQt4.QtGui import *

from pcal import *
from kpi_dicts import *

# import signal # TODO add time limit

import socket
socket.setdefaulttimeout(10)

img_path = '/usr/share/pixmaps/kpi/'
# img_path = './img/'
# d = base64.b64encode(bytes(domain_url, "utf-8"))
domain_url = base64.b64decode(domain_url).decode("utf-8", "ignore")

class AppSettings():
	def __init__(self, parent=None):
		super(AppSettings, self).__init__()
		self.settings = QSettings()
		self.authstring = self.settings.value('authstring')

	def getParametr(self, name):
		if hasattr(self, name):
			attr = getattr(self, name)
			if attr != None:
				return attr
		attr = self.settings.value(name)
		if attr != None:
			return attr
		else:
			return False

	def setParametr(self, name, value):
		setattr(self, name, value)
		self.settings.setValue(name, value)
		return True

class GetKPI():
	def __init__(self, parent=None):
		super(GetKPI, self).__init__(parent)
		# add init code here

	def get_auth_url(url, username, password):
		# TODO: implement QNetworkRequest and QAbstractNetworkCache
		data = urllib.parse.urlencode({'ldap-mail': username, 'ldap-pass': password, 'go': ' Войти '})
		data = data.encode('ascii') # data should be bytes
		request = urllib.request.Request(url, data)
		resorce = urllib.request.urlopen(request)
		html = resorce.read().decode("utf-8").strip()
		return html

	def auth_probe(username, password):
		content = GetKPI.get_auth_url(domain_url + api_uri + api_employees_get, username, password)
		if content.find("Авторизация LDAP") > 0:
			return False
		return True

class MyWindow(QWidget):
	def __init__(self, data_list, header, userpos, *args):
		QWidget.__init__(self, *args)
		# setGeometry(x_pos, y_pos, width, height)
		self.setGeometry(300, 200, 570, 450)
		self.setWindowTitle("Demis KPI")
		self.kpi = ''
		self.setWindowIcon(QIcon(img_path+'app.png'))
		table_model = MyTableModel(self, data_list, header)
		self.table_view = QTableView()
		self.table_view.setModel(table_model)
		self.sett = AppSettings()

		self.name_vacation = 'vacation'
		self.name_prediction = 'prediction'
		self.name_vip = 'vip'
		self.name_group = 'group'

		prediction_list = []
		vacation_list = []
		vip_list = []
		group_list = []
		c = 0
		for h in header:
			if c != 0:
				if re.search('Прогноз', h):
					prediction_list.append(c)
				if h.find('тсутстви') > 0:
					vacation_list.append(c)
				if re.search('\sVIP', h):
					vip_list.append(c)
				if re.search('подразделение', h, re.IGNORECASE):
					group_list.append(c)
			c = c + 1

		# http://doc.qt.io/qt-5/stylesheet-examples.html
		# http://doc.qt.io/qt-5/stylesheet-reference.html
		style = '''border: none;'''
		# отключаем показ границ виджета
		self.table_view.setStyleSheet(style)

		# Установка шрифта
		self.setStyleSheet('''/* font: 12pt "Arial"; */

		QHeaderView::up-arrow, QHeaderView::down-arrow, QHeaderView::indicator, QTableView::indicator {color:#fff;
		margin: 0;
		padding: 0;
		spacing: 0;
		margin-right: -10px;
		width: 2px;
		max-width: 2px;
		/* subcontrol-origin: content;
		subcontrol-origin: margin;
		subcontrol-origin: border; */
		subcontrol-position: right center;
		position: absolute;
		right: -10px;
		icon-size: 2px;
		show-decoration-selected: 0;
		opacity: 0;
		viasbility: hidden;
		font-size: 1px;
		outline-offset:0;
		border: none;
		background: none;
		}

		QTableView::item:hover {
			/* background: qlineargradient(x1: 0, y1: 0, x2: 0, y2: 1, stop: 0 #cfcfdd, stop: 1 #f0f0f0); */
			background: grey;
			border: 1px solid #000;
		}
			''')
		font = QFont("Arial", 12)
		self.table_view.setFont(font)

		# set column width to fit contents (set font first!)
		self.table_view.resizeColumnsToContents()

		# set selection mode
		# self.table_view.setSelectionBehavior(QAbstractItemView.SelectRows)
		
		# пользователь не может изменять размер столбцов
		# self.table_view.horizontalHeader().setResizeMode(QHeaderView.Fixed)

		self.table_view.horizontalHeader().setCascadingSectionResizes(False)
		# self.table_view.horizontalHeader().setDefaultSectionSize(100)

		# включаем растягивание последнего столбца
		self.table_view.horizontalHeader().setStretchLastSection(False)

		# Сетка
		# self.table_view.setShowGrid(False)

		# Включаем чередующуюся подсветку строк
		self.table_view.setAlternatingRowColors(True)

		# скрываем заголовки строк
		# self.table_view.verticalHeader().hide()

		# enable sorting
		self.table_view.setSortingEnabled(True)
		self.table_view.selectRow(userpos)
		self.table_view.sortByColumn(0, Qt.AscendingOrder)
		layout = QVBoxLayout(self)

		self.table_view.hideColumn(0)
		
		self.statusbar = QStatusBar()
		self.statusbar.setObjectName("statusbar")

		self.myQMenuBar = QMenuBar(self)
		fileMenu = self.myQMenuBar.addMenu('File')

		exitAction = QAction(QIcon(img_path+'exit.png'), 'Закрыть', self)
		exitAction.setShortcut('esc')      
		exitAction.triggered.connect(qApp.quit)
		fileMenu.addAction(exitAction)

		# fileMenu.addSeparator()

		restartAction = QAction(QIcon(img_path+'reload.png'), 'Обновить', self)
		restartAction.setShortcut('f5')       
		restartAction.triggered.connect(self.action_reload)
		fileMenu.addAction(restartAction)

		fileMenu = self.myQMenuBar.addMenu('Preferences')

		set_id_Action = QAction('Настройки', self)
		set_id_Action.setShortcut('alt+i')       
		set_id_Action.triggered.connect(self.action_set_settings)
		fileMenu.addAction(set_id_Action)

		toggleVIPAction = QAction(QIcon(img_path+'vip.png'), 'VIP', self)
		toggleVIPAction.setShortcut('alt+v')
		toggleVIPAction.setCheckable(True)
		# toggleVIPAction.setChecked(True)
		toggleVIPAction.triggered.connect(lambda: self.toggle_colls([self.name_vip, 'VIP'], vip_list))

		toggleGroupAction = QAction(QIcon(img_path+'group.png'), 'Показатели подразделений', self)
		toggleGroupAction.setShortcut('alt+g')
		toggleGroupAction.setCheckable(True)
		# toggleGroupAction.setChecked(True)
		toggleGroupAction.triggered.connect(lambda: self.toggle_colls([self.name_group, 'показатели подразделений'], group_list))

		toggleVacationAction = QAction(QIcon(img_path+'vacation.png'), 'Отпуск и отсутствия', self)
		toggleVacationAction.setShortcut('alt+w')
		toggleVacationAction.setCheckable(True)
		# toggleVacationAction.setChecked(True)
		toggleVacationAction.triggered.connect(lambda: self.toggle_colls([self.name_vacation, 'отсутствия'], vacation_list))

		togglePredictionAction = QAction(QIcon(img_path+'prediction.png'), 'Прогноз', self)
		togglePredictionAction.setShortcut('alt+p')
		togglePredictionAction.setCheckable(True)
		# togglePredictionAction.setChecked(True)
		togglePredictionAction.triggered.connect(lambda: self.toggle_colls([self.name_prediction, 'прогноз'], prediction_list))

		# copyAction = QAction(self)
		# copyAction.setShortcut('ctrl+ins')
		# copyAction.setShortcut('ctrl+c')
		# copyAction.setShortcut(QKeySequence(Qt.Key_C))
		# copyAction.triggered.connect(self.copy_cells_to_clipboard)
		# self.addAction(copyAction)

		self.toolbar = QToolBar(self)
		self.toolbar.addAction(exitAction)
		self.toolbar.addAction(restartAction)
		# self.toolbar.addSeparator()
		self.toolbar.addAction(toggleVIPAction)
		self.toolbar.addAction(toggleGroupAction)
		self.toolbar.addAction(toggleVacationAction)
		self.toolbar.addAction(togglePredictionAction)

		self.setLayout(layout)
		layout.addWidget(self.myQMenuBar)
		layout.addWidget(self.toolbar)
		layout.addWidget(self.table_view)
		layout.addWidget(self.statusbar)

		if self.sett.getParametr(self.name_vip) == '1' or self.sett.getParametr(self.name_vip) == 1:
			self.toggle_colls([self.name_vip, 'VIP'], vip_list, False)
			toggleVIPAction.setChecked(True)

		if self.sett.getParametr(self.name_group) == '1' or self.sett.getParametr(self.name_group) == 1:
			self.toggle_colls([self.name_group, 'показатели подразделений'], group_list, False)
			toggleGroupAction.setChecked(True)

		if self.sett.getParametr(self.name_prediction) == '1' or self.sett.getParametr(self.name_prediction) == 1:
			self.toggle_colls([self.name_prediction, 'прогноз'], prediction_list, False)
			togglePredictionAction.setChecked(True)

		if self.sett.getParametr(self.name_vacation) == '1' or self.sett.getParametr(self.name_vacation) == 1:
			self.toggle_colls([self.name_vacation, 'отсутствия'], vacation_list, False)
			toggleVacationAction.setChecked(True)

	def keyPressEvent(self, e):
		if (e.modifiers() & Qt.ControlModifier):
			if e.key() in (Qt.Key_C, 16777222): # QKeySequence.Copy
				self.copy_cells_to_clipboard()

	def copy_cells_to_clipboard(self):
		if len(self.table_view.selectionModel().selectedIndexes()) > 0:
			# sort select indexes into rows and columns
			previous = self.table_view.selectionModel().selectedIndexes()[0]
			columns = []
			rows = []
			for index in self.table_view.selectionModel().selectedIndexes():
				if previous.column() != index.column():
					columns.append(rows)
					rows = []
				rows.append(index.data())
				previous = index
			columns.append(rows)

			# add rows and columns to clipboard            
			clipboard = ""
			nrows = len(columns[0])
			ncols = len(columns)
			for r in range(nrows):
				for c in range(ncols):
					clipboard += str(columns[c][r]).replace('.', ',')
					if c != (ncols-1):
						clipboard += '\t'
				clipboard += '\n'

			# copy to the system clipboard
			sys_clip = QApplication.clipboard()
			sys_clip.setText(clipboard)

	# def closeEvent(self,event):
	# 	reply = QMessageBox.question(self,'Message',"Are you sure to quit?", QMessageBox.Yes, QMessageBox.No)
	# 	if reply == QMessageBox.Yes:
	# 		event.accept()
	# 	else:
	# 		event.ignore()

	def event(self, event):
		if event.type() == QEvent.WindowStateChange:
			if self.isMinimized():
				# self.setWindowFlags(self.windowFlags() & ~Qt.Tool)
				self.hide()
				self.sett.setParametr("window_state", 'minimized');
		else:
			return super(MyWindow, self).event(event)
		return True

	def scrollTo(self, index, hint):
		if index.column() > 1:
			QTableView.scrollTo(self, index, hint)

	def toggle_colls(self, colls_type, cols_list, mode=True):
		# TODO store ranges in app settings
		if len(cols_list) > 0:
			if self.sett.getParametr(colls_type[0]) == '1' or self.sett.getParametr(colls_type[0]) == 1:
				for i in cols_list:
					if mode: self.table_view.showColumn(i)
					else: self.table_view.hideColumn(i)
				if mode: self.sett.setParametr(colls_type[0], 0)
			else:
				for i in cols_list:
					if mode: self.table_view.hideColumn(i)
					else: self.table_view.showColumn(i)
				if mode: self.sett.setParametr(colls_type[0], 1)
			self.statusbar.showMessage('изменение видимости - '+colls_type[1])
		return

	def action_reload(self):
		# TODO: reload model only - http://www.qtcentre.org/threads/31770-How-to-inform-a-Model-TableView-that-some-data-changed
		subprocess.Popen([__file__])
		sys.exit(0)

	def action_set_settings(self):
		saver = ChangeSettings()
		if saver.exec_() == QDialog.Accepted:
			self.statusbar.showMessage('Данные сохранены')
			self.action_reload()
		else:
			self.statusbar.showMessage('Ошибка сохранения данных!')
		return

class MyTableModel(QAbstractTableModel):
	def __init__(self, parent, mylist, header, *args):
		QAbstractTableModel.__init__(self, parent, *args)
		self.mylist = mylist
		self.header = header

	def rowCount(self, parent):
		return len(self.mylist)

	def columnCount(self, parent):
		return len(self.mylist[0])

	def data(self, index, role):
		if not index.isValid():
			return None
		elif role != Qt.DisplayRole:
			return None
		return self.mylist[index.row()][index.column()]

	def headerData(self, section, orientation, role):
		if orientation == Qt.Horizontal and role == Qt.DisplayRole:
			return self.header[section]
		elif orientation == Qt.Vertical:
		# 	return ['Row %d' % int(row+1) for row in range(len(self.mylist))]
			index = self.index(section, 0) # assuming ID is the first column
			return self.data(index, role)
		
		return QAbstractTableModel.headerData(self, section, orientation, role)
		# return None

	def sort(self, col, order):
		"""sort table by given column number col"""
		self.emit(SIGNAL("layoutAboutToBeChanged()"))
		self.mylist = sorted(self.mylist,
			key=operator.itemgetter(col))
		if order == Qt.DescendingOrder:
			self.mylist.reverse()
		self.emit(SIGNAL("layoutChanged()"))

class ChangeSettings(QDialog):
	def __init__(self, parent=None):
		super(ChangeSettings, self).__init__(parent)
		self.sett = AppSettings()

		self.loginLabel = QLabel("Login: ")
		self.textName = QLineEdit(self)
		self.textName.setText(self.sett.getParametr("username"))

		self.passwordLabel = QLabel("Password: ")
		self.textPass = QLineEdit(self)
		self.textPass.setText(self.sett.getParametr("password"))

		self.periodLabel = QLabel("Период обновления: ")
		self.listWidget = QListWidget()
		for i in ('не обновлять','15','30','60'):
			item = QListWidgetItem(i)
			self.listWidget.addItem(item)
		refresh_period = self.sett.getParametr("refresh_period")
		if refresh_period:
			for i in self.listWidget.findItems(refresh_period, Qt.MatchFixedString):
				if i.text() == refresh_period:
					self.listWidget.setCurrentItem(i)
					break

		self.buttonSave = QPushButton('Сохранить', self)
		self.buttonSave.clicked.connect(self.handleSubmit)

		layout = QVBoxLayout(self)
		layout.addWidget(self.loginLabel)
		layout.addWidget(self.textName)
		layout.addWidget(self.passwordLabel)
		layout.addWidget(self.textPass)
		layout.addWidget(self.periodLabel)
		layout.addWidget(self.listWidget)
		layout.addWidget(self.buttonSave)

	def handleSubmit(self):
		self.sett.setParametr("username", self.textName.text())
		self.sett.setParametr("password", self.textPass.text())
		# if self.listWidget.selectedItems()[0].text().find('обновлять') < 1:
		if len(self.listWidget.selectedItems()[0].text()) < 3:
			refresh_period = self.sett.setParametr("refresh_period", self.listWidget.selectedItems()[0].text())
		else:
			refresh_period = self.sett.setParametr("refresh_period", '0')
		self.accept()

class Login(QDialog):
	def __init__(self, parent=None):
		super(Login, self).__init__(parent)
		self.textName = QLineEdit(self)
		self.textName.setText("login")
		self.textPass = QLineEdit(self)
		self.textPass.setText("password")
		self.buttonLogin = QPushButton('Login', self)
		self.buttonLogin.clicked.connect(self.handleLogin)
		layout = QVBoxLayout(self)
		layout.addWidget(self.textName)
		layout.addWidget(self.textPass)
		layout.addWidget(self.buttonLogin)

	def handleLogin(self):
		sett = AppSettings()
		sett.setParametr("username", self.textName.text())
		sett.setParametr("password", self.textPass.text())

		if GetKPI.auth_probe(self.textName.text(), self.textPass.text()) != False:
			self.accept()
		else:
			QMessageBox.warning(
				self, 'Error', 'Неверный логин или пароль!')

class SystemTrayIcon(QSystemTrayIcon):
	def __init__(self, icon, parent, kpi):
		QSystemTrayIcon.__init__(self, icon, parent)
		self.win = parent
		self.kpi = kpi
		menu = QMenu(parent)

		showAction = QAction('Показать окно программы', self)
		# showAction.setShortcut('ctrl+h')      
		showAction.triggered.connect(self.show_action)
		menu.addAction(showAction)

		exitAction = QAction('Exit', self)
		exitAction.setShortcut('esc')      
		exitAction.triggered.connect(qApp.quit)
		menu.addAction(exitAction)

		self.setContextMenu(menu)

		# on activation listener
		self.activated.connect(self.trayActivated)

	def show_action(self):
		self.win.setWindowState(Qt.WindowNoState)
		self.win.show()
		# self.win.setWindowState(Qt.WindowActive)
		AppSettings.setParametr(self.win.sett, "window_state", 'shown');
		return True

	def trayActivated(self, reason):
		if reason == QSystemTrayIcon.DoubleClick:
			self.show_action()
		elif reason == QSystemTrayIcon.Trigger:
			CommonTools.show_popup('всего баллов', str(self.kpi))    

	def create_large_icon(self, text):
		font = QFont("Arial", 24)
		font.setPointSizeF(font.pointSizeF() * 2)
		font.setWeight(600)
		metrics = QFontMetricsF(font)

		rect = metrics.boundingRect(text)
		position = -rect.topLeft()
		
		if not text:
				return
		
		pixmap = QPixmap(rect.width(), rect.height())
		pixmap.fill(Qt.white)
		
		painter = QPainter()
		painter.begin(pixmap)
		painter.setFont(font)
		painter.setPen(QColor(0, 0, 0))
		# painter.drawText(1, 1, 64, 64, Qt.AlignLeft, text)
		painter.drawText(position, text)
		painter.end()
	
		return(pixmap)

	def create_icon(self, text):
		if not text: return
		font = QFont("Arial", 26)
		font.setWeight(600)
		pixmap = QPixmap(64, 64)
		pixmap.fill(Qt.white)
		
		painter = QPainter()
		painter.begin(pixmap)
		painter.setFont(font)
		painter.setPen(QColor(0, 0, 0))
		painter.drawText(1, 14, 64, 64, Qt.AlignLeft, text)
		painter.end()
		return(pixmap)

class CommonTools():
	def show_popup(header, data):
		cmd = "notify-send '" + str(header) + "' '" + str(data) + "' '-t' 5000"
		process = subprocess.Popen(cmd, shell=True, stdin=None, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True)
		# process = subprocess.call(['notify-send', data])

		# stdout, stderr = process.communicate() # only with Popen
		# print(stdout)
		# print(stderr)
		return True

	def closeEvent(window):
		sett.setParametr("geometry", window.saveGeometry());
		# sett.setParametr("windowState", app.saveState());
		return True

if __name__=="__main__":

	def split_header(text):
		text = text.split(' ', 2)
		s = "\n";
		text =  s.join( text )
		return text

	app = QApplication([]) # app = QApplication(sys.argv)
	sett = AppSettings()
	auth = False
	header = ['Сотрудник']
	icon_data = ''
	userpos = 0
	users = []

	username = sett.getParametr("username")
	password = sett.getParametr("password")
	# username = sett.setParametr("username", "")
	# password = sett.setParametr("password", "")

	if GetKPI.auth_probe(username, password) == False:
		print("Требуется авторизация")
		login = Login()
		if login.exec_() == QDialog.Accepted:
			auth = True
			subprocess.Popen([__file__])
			sys.exit(0)

	else:
		auth = True

	if auth == True:

		full_url = domain_url + api_uri + api_result_get
		rjson = GetKPI.get_auth_url(full_url, username, password)

		try:
			cdict = json.loads(rjson)
		except:
			raise ValueError('Сервер вернул пустой ответ')

		full_url = domain_url + api_uri + api_employees_get
		rjson = GetKPI.get_auth_url(full_url, username, password)
		
		try:
			udict = json.loads(rjson)
		except:
			raise ValueError('Сервер вернул пустой ответ')

		data_list = []

		for record in cdict:
			user_data = ()

			emp_name = str(udict[record]['name'])
			grade = str(udict[record]['grade_name'])

			users += [emp_name,]

			user_data += (emp_name + " (" + grade + ")",)

			if (udict[record]['login'] == username) and (icon_data == ''):
				icon_data = str(cdict[record]['result'])
				notify_name = emp_name

			od = collections.OrderedDict(sorted(cdict[record].items(), reverse=True))
			for r_feild in od:
				try:
					indicator_name = split_header(str(result_rus_dict[r_feild]))
				except KeyError as e:
					indicator_name = str(r_feild)
					# raise ValueError('Undefined unit: {}'.format(e.args[0]))

				if len(header) <= len(od):
					header = header + [indicator_name,]

				indicator = round(float(cdict[record][r_feild]),2)
				# user_data += (str(indicator) + " (" + indicator_name + ")",)
				user_data += (indicator,)
			data_list = data_list + [user_data,]
		
		userpos = sorted(users).index(notify_name)

		win = MyWindow(data_list, header, userpos)
		app.aboutToQuit.connect(lambda: CommonTools.closeEvent(win))

		trayIcon = SystemTrayIcon(QIcon(img_path+"app.png"), win, icon_data)
		trayIcon.setIcon(QIcon(trayIcon.create_icon(icon_data)))

		trayIcon.setToolTip(icon_data)
		trayIcon.show()

		win.setMinimumSize(800, 600)
		win.resize(1024, 768)

		last_geom = sett.getParametr("geometry")
		window_state = sett.getParametr("window_state")
		refresh_period = int(sett.getParametr("refresh_period"))

		if last_geom:
			if type(last_geom) is not bytearray:
				last_geom = bytearray()
				last_geom.extend(sett.getParametr("geometry"))
			win.restoreGeometry(last_geom)
		if window_state == 'minimized':
			win.setWindowState(Qt.WindowMinimized)
		else:
			win.show()
		win.statusbar.showMessage('Ready')
		CommonTools.show_popup(notify_name +' - всего баллов', icon_data)

		if refresh_period and refresh_period>0:
			refresh_period = refresh_period*60000
			timer = QTimer()
			timer.timeout.connect(lambda: win.action_reload())
			timer.start(refresh_period) # 60000 trigger every minute.
		
		app.exec_()

	else:
		app.quit
		


opt/kpi/pcal.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

'''Производственный календарь'''

import calendar, datetime

class pCal:

	#http://data.gov.ru/opendata/7708660670-proizvcalendar
	p_clndr =  {
	'2016':[
	[1,2,3,4,5,6,7,8,9,10,16,17,23,24,30,31],
	[6,7,13,14,21,22,23,27,28],
	[5,6,7,8,12,13,19,20,26,27],
	[2,3,9,10,16,17,23,24,30],
	[1,2,3,7,8,9,14,15,21,22,28,29],
	[4,5,11,12,13,18,19,25,26],
	[2,3,9,10,16,17,23,24,30,31],
	[6,7,13,14,20,21,27,28],
	[3,4,10,11,17,18,24,25],
	[1,2,8,9,15,16,22,23,29,30],
	[4,5,6,12,13,19,20,26,27],
	[3,4,10,11,17,18,24,25,31]
	],
	'2017':[
	[1,2,3,4,5,6,7,8,9,14,15,21,22,28,29],
	[4,5,11,12,18,19,23,25,26],
	[4,5,8,11,12,18,19,25,26],
	[1,2,8,9,15,16,22,23,29,30],
	[1,6,7,8,9,13,14,20,21,27,28],
	[3,4,10,11,12,17,18,24,25],
	[1,2,8,9,15,16,22,23,29,30],
	[5,6,12,13,19,20,26,27],
	[2,3,9,10,16,17,23,24,30],
	[1,7,8,14,15,21,22,28,29],
	[4,5,6,11,12,18,19,25,26],
	[2,3,9,10,16,17,23,24,30,31]
	],
	'2018':[
	[1,2,3,4,5,6,7,8,9,10,13,14,20,21,27,28],
	[3,4,10,11,17,18,23,24,25],
	[3,4,8,10,11,17,18,24,25,31],
	[1,7,8,14,15,21,22,28,29],
	[1,5,6,9,12,13,19,20,26,27],
	[2,3,9,10,12,16,17,23,24,30],
	[1,7,8,14,15,21,22,28,29],
	[4,5,11,12,18,19,25,26],
	[1,2,8,9,15,16,22,23,29,30],
	[6,7,13,14,20,21,27,28],
	[3,4,5,10,11,17,18,24,25],
	[1,2,8,9,15,16,22,23,29,30]
	],
	'2019':[
	[1,2,3,4,5,6,7,8,9,10,12,13,19,20,26,27],
	[2,3,9,10,16,17,23,24,25],
	[2,3,8,9,10,16,17,23,24,30,31],
	[6,7,13,14,20,21,27,28],
	[4,5,9,11,12,18,19,25,26],
	[1,2,8,9,12,15,16,22,23,29,30],
	[6,7,13,14,20,21,27,28],
	[3,4,10,11,17,18,24,25,31],
	[1,7,8,14,15,21,22,28,29],
	[5,6,12,13,19,20,26,27],
	[2,3,9,10,16,17,23,24,30],
	[1,7,8,14,15,21,22,28,29]
	]
	}

	now = datetime.datetime.now()

	def month_days(month_number='current'):
		if month_number == 'current':
			return calendar.monthrange(pCal.now.year, pCal.now.month-1)[1]
		elif month_number == 'previous':
			return calendar.monthrange(pCal.now.year, pCal.now.month-2)[1]
		elif isinstance(month_number, int) and (int(month_number) > 0) and (int(month_number) < 13):
			return calendar.monthrange(pCal.now.year, int(month_number))[1]
		else:
			return False

	def working_days(month_number='current'):
		if month_number == 'current':
			return str( pCal.month_days() - len(pCal.p_clndr[str(pCal.now.year)][pCal.now.month-1]) )
		elif month_number == 'previous':
			return str( pCal.month_days(pCal.now.month-2) - len(pCal.p_clndr[str(pCal.now.year)][pCal.now.month-2]) )
		elif isinstance(month_number, int) and (int(month_number) > 0) and (int(month_number) < 13):
			return str( pCal.month_days(int(month_number)) - len(pCal.p_clndr[str(pCal.now.year)][int(month_number)]) )
		else:
			return False

	def working_days_passed(day='current'):
		total_holidays = 0

		if day == 'current':
			to_day = int(pCal.now.day)
		elif int(day) in range(1,31):
			to_day = int(day)
		else:
			return False

		# Считать сегодня или нет
		# if(pCal.now.hour < 14):
		# 	to_day = to_day - 1

		for d in pCal.p_clndr[str(pCal.now.year)][pCal.now.month-1]:
			if int(d) < to_day:
				total_holidays = total_holidays + 1
			else:
				break
		return to_day - total_holidays

